smithers/ml/fnn.py

```diff
@@ -0,0 +1,107 @@
+        self.fc3 = nn.Linear(self.n_hid, self.n_hid)
+        self.fc4 = nn.Linear(self.n_hid, self.n_hid)
+        self.fc5 = nn.Linear(self.n_hid, self.n_hid)
+        self.fc6 = nn.Linear(self.n_hid, self.n_hid)
+        self.fc7 = nn.Linear(self.n_hid, self.n_class)
+        self.cifar = cifar
+
+    def forward(self, x):
+        '''
+        Forward Phase.
+
+        :param tensor x: input of the network with dimensions
+            n_images x n_input
+        :return: output of the FNN n_images x n_class
+        :rtype: tensor
+        '''
+        if self.cifar is not None:
+            x = x.view(x.size(0), -1)
+#        x = torch.nn.Softplus()(self.fc1(x))
+#        x = torch.nn.Softplus()(self.fc2(x))
+        x = torch.nn.ReLU()(self.fc1(x))
+        x = torch.nn.ReLU()(self.fc2(x))
+        x = torch.nn.ReLU()(self.fc3(x))
+        x = torch.nn.ReLU()(self.fc4(x))
+        x = torch.nn.ReLU()(self.fc5(x))
+        x = torch.nn.ReLU()(self.fc6(x))
+        x = self.fc7(x)
+        return x
+
+
+
+def training_fnn(fnn_net, epochs, inputs_net, real_out):
+    '''
+    Training phase for a Feed Forward Neural Network (FNN).
+
+    :param nn.Module fnn_net: FNN model
+    :param int epochs: epochs for the training phase.
+    :param tensor inputs_net: matrix of inputs for the network
+        with dimensions n_input x n_images.
+    :param tensor real_out: tensor representing the real output
+        of the network.
+    '''
+    criterion = nn.CrossEntropyLoss()
+    optimizer = optim.Adam(fnn_net.parameters(), lr=0.001)
+    correct = 0
+    total = 0
+
+    final_loss = []
+    batch_size = 128
+    for epoch in range(epochs):  # loop over the dataset multiple times
+        running_loss = 0.0
+        for i in range(inputs_net.size()[0] // batch_size):
+        # zero the parameter gradients
+            optimizer.zero_grad()
+
+            # forward + backward + optimize
+            outputs = fnn_net(inputs_net[i * batch_size:(i + 1) *
+                                         batch_size, :])
+            loss = criterion(
+                outputs,
+                torch.LongTensor(real_out[i * batch_size:(i + 1) *
+                                          batch_size]))
+            loss.backward(retain_graph=True)
+            optimizer.step()
+
+            # print statistics
+            running_loss += loss.item()
+            if i % 50 == 49:  # print every 500 mini-batches
+                print('[%d, %5d] loss: %.6f' %
+                      (epoch + 1, i + 1, running_loss / 50))
+                if i == inputs_net.size()[0] // batch_size - 1:
+                    final_loss.append(running_loss / 50)
+                running_loss = 0.0
+
+            _, predicted = torch.max(outputs.data, 1)
+            labels = torch.LongTensor(real_out[i * batch_size:(i + 1) *
+                                               batch_size])
+            total += labels.size(0)
+            correct += (predicted == labels).sum().item()
```

Comment:
I would keep the FNN customizable by user (right now the architecture is hardcoded). Maybe we can use the model from [PINA](https://github.com/mathLab/PINA/blob/main/pina/model/feed_forward.py)?

Reply:
Ok. Should I import that class from PINA?
I think I have only to check that everything is working for the CIFAR dataset, where I have placed in the forward `x = x.view(x.size(0), -1)` in order to have everything working correctly. 
Then the function training_fnn should be removed and placed in the utils.py file or has already been implemented in PINA?

Reply:
Yes, for now I would just import directly from PINA (in future we can evaluate other possibilities)